esphome/components/uart/uart_esp8266.cpp

```diff
@@ -49,15 +55,29 @@ void UARTComponent::setup() {
   // is 1 we still want to use Serial.
   SerialConfig config = static_cast<SerialConfig>(get_config());
 
-  if (this->tx_pin_.value_or(1) == 1 && this->rx_pin_.value_or(3) == 3) {
+  if (!esphome::uart::UARTComponent::serial0InUse && this->tx_pin_.value_or(1) == 1 &&
+      this->rx_pin_.value_or(3) == 3
+#ifdef USE_LOGGER
+      // we will use UART0 if Logger is using either UART0 or UART1 or NONE
+      && !(logger::global_logger->get_hw_serial() == &Serial && (IOSWAP & (1 << IOSWAPU0)))
```

Comment:
We don't need to poke around in internals to detect swap state here, we can just use the `get_uart()` method of the logger:
```suggestion
      // we will use UART0 if logger isn't using it in swapped mode
      && logger::global_logger->get_uart() != logger::UART_SELECTION_UART0_SWAP
```